llarp/config/config.cpp

```diff
@@ -936,48 +933,95 @@ namespace llarp
       {
-        // infer public address
-        if (auto maybe_ifname = net_ptr->GetBestNetIF())
-          addr = net_ptr->GetInterfaceAddr(*maybe_ifname);
+        uint16_t res = std::atoi(arg_v.substr(1).data());
+
+        maybe = oxen::quic::Address{""s, res ? res : DEFAULT_LISTEN_PORT};
       }
+      else if (auto pos = arg_v.find(':'); pos != arg_v.npos)
+      {
+        auto h = arg_v.substr(0, pos);
+        uint16_t p = std::atoi(arg_v.substr(pos + 1).data());
+
+        maybe = oxen::quic::Address{std::string{h}, p};
 
-      if (addr)
+        // TODO: unfuck llarp/net
+        // if (net_ptr->IsLoopbackAddress(addr->port()))
+        //   throw std::invalid_argument{fmt::format("{} is a loopback address", arg)};
+      }
+      if (not maybe)
       {
-        // set port if not explicitly provided
-        if (addr->getPort() == 0)
-          addr->setPort(default_port);
+        // infer public address
+        if (auto maybe_ifname = net_ptr->GetBestNetIF())
+          maybe = oxen::quic::Address{*maybe_ifname};
       }
-      return addr;
+
+      if (maybe->port() == 0)
+        maybe = oxen::quic::Address{maybe->host(), DEFAULT_LISTEN_PORT};
+
+      return maybe;
     };
 
+    conf.define_option<std::string>(
+        "bind",
+        "listen",
+        Required,
+        Comment{
+            "********** NEW API OPTION (see note) **********",
+            "",
+            "IP and/or port for lokinet to bind to for inbound/outbound connections.",
+            "",
+            "If IP is omitted then lokinet will search for a local network interface with a",
+            "public IP address and use that IP (and will exit with an error if no such IP is found",
+            "on the system).  If port is omitted then lokinet defaults to 1090.",
+            "",
+            "Note: only one address will be accepted. If this option is not specified, it will "
+            "default",
+            "to the inbound or outbound value. Conversely, specifying this option will supercede "
+            "the",
+            "deprecated inbound/outbound opts.",
+            "",
+            "Examples:",
+            "    listen=15.5.29.5:443",
+            "    listen=10.0.2.2",
+            "    listen=:1234",
+            "",
+            "Using a private range IP address (like the second example entry) will require using",
+            "the public-ip= and public-port= to specify the public IP address at which this",
+            "router can be reached.",
+        },
+        [this, parse_addr_for_link](const std::string& arg) {
+          if (auto a = parse_addr_for_link(arg); a and a->is_addressable())
+            addr = a;
+          else
+            addr = oxen::quic::Address{""s, DEFAULT_LISTEN_PORT};
+
+          using_new_api = true;
+        });
+
     conf.define_option<std::string>(
         "bind",
         "inbound",
         RelayOnly,
         MultiValue,
         Comment{
+            "********** DEPRECATED **********",
```

Comment:
Instead of a comment like this, we should just add the `Deprecated` flag to the option definition, so that in generated configs (`lokinet -c -g dir/lokinet.ini` or `lokinet -r -g dir/lokinet.ini` which is what generates a new default config) the section won't show up at all.

Reply:
Whoops, I wrong: we want `Hidden`, not `Deprecated` to hide it but still process it.  (`Deprecated` is for a no-longer-does-anything-at-all option that we want to just ignore but not error on if we see it).

Reply:
I didn't add a deprecated flag to the function parameters? I just put deprecated as the first comment line. I added "Hidden" after "RelayOnly" and "MultiValue," is that what you meant?